ecommerce/extensions/payment/views/payu.py

```diff
@@ -0,0 +1,241 @@
+Basket = get_model('basket', 'Basket')
+BillingAddress = get_model('order', 'BillingAddress')
+Country = get_model('address', 'Country')
+NoShippingRequired = get_class('shipping.methods', 'NoShippingRequired')
+OrderNumberGenerator = get_class('order.utils', 'OrderNumberGenerator')
+OrderTotalCalculator = get_class('checkout.calculators', 'OrderTotalCalculator')
+PaymentProcessorResponse = get_model('payment', 'PaymentProcessorResponse')
+
+
+class PayuPaymentExecutionView(EdxOrderPlacementMixin, View):
+    """Execute an approved PayU payment and place an order for paid products as appropriate."""
+    @property
+    def payment_processor(self):
+        return PayU(self.request.site)
+
+    @method_decorator(transaction.non_atomic_requests)
+    def dispatch(self, request, *args, **kwargs):
+        return super(PayuPaymentExecutionView, self).dispatch(request, *args, **kwargs)
+
+    def _get_basket(self, payment_id):
+        """
+        Retrieve a basket using a payment ID.
+
+        Arguments:
+            payment_id: payment_id received from PayPal.
+
+        Returns:
+            It will return related basket or log exception and return None if
+            duplicate payment_id received or any other exception occurred.
+
+        """
+        try:
+            basket = PaymentProcessorResponse.objects.get(
+                processor_name=self.payment_processor.NAME,
+                transaction_id=payment_id
+            ).basket
+            basket.strategy = strategy.Default()
+            Applicator().apply(basket, basket.owner, self.request)
+            return basket
+        except MultipleObjectsReturned:
+            logger.exception(u"Duplicate payment ID [%s] received from PayPal.", payment_id)
+            return None
+        except Exception:  # pylint: disable=broad-except
+            logger.exception(u"Unexpected error during basket retrieval while executing PayPal payment.")
+            return None
+
+    def get(self, request):
+        """Handle an incoming user returned to us by PayPal after approving payment."""
+        payment_id = request.GET.get('orderNum')
+        transactionState = request.GET.get('transactionState')
+        logger.info(u"PayU payment [%s] transactionState [%s]", payment_id, transactionState)
+
+        payu_response = request.GET.dict()
+        basket = self._get_basket(payment_id)
+
+        if not basket:
+            return redirect(self.payment_processor.error_url)
+
+        receipt_url = get_receipt_page_url(
+            order_number=basket.order_number,
+            site_configuration=basket.site.siteconfiguration
+        )
+
+        try:
+            with transaction.atomic():
+                try:
+                    self.handle_payment(payu_response, basket)
+                except PaymentError:
+                    logger.exception('PaymentError for basket [%d] failed.', basket.id)
+                    if transactionState == '7':
```

Comment:
This is fucking terrible, can you add a comment above this line as a TODO to investigate why the value '7'?

Reply:
in the previous comments said that it meant transactionState